### ISSUE TITLE
Unexpected content of "related" link

### ISSUE THREAD

**Issue by david-kubecka** [crnk-project/crnk-framework#611]
I'm currently familiarizing myself with CRNK and have run into issue (or misunderstanding) regarding the response on GET on resource's "related" link. Here's my setup (only important bits included for clarity). Basically two objects, backed by provided in-memory repository with two-way relationship between them (with default lookUp, repositoryBehavior etc. parameters):
```
@JsonApiResource(type = "attributes")
public class Attribute {
    @JsonApiId
    private Long id;

    @JsonApiRelation(mappedBy="attribute")
    private List<Label> labels;
}

@JsonApiResource(type = "labels")
public class Label {
    @JsonApiId
    private Long id;

    @JsonApiRelation
    private Attribute attribute;
}

@Component
public class AttributeRepositoryImpl extends InMemoryResourceRepository<Attribute, Long> {
    public AttributeRepositoryImpl() {
        super(Attribute.class);
    }
}

@Component
public class LabelRepositoryImpl extends InMemoryResourceRepository<Label, Long> {
    public LabelRepositoryImpl() {
        super(Label.class);
    }
}
```
Now I create two objects, one for each type, as follows:
```
{
  "data": {
    "id": 9,
    "type": "labels"
  }
}
```
```
{
  "data": {
    "id": 123,
    "type": "attributes",
    "relationships": {
      "labels": {
        "data": [{ "type": "labels", "id": "9" }]
      }
    }
  }
}
```
Next I query the created attribute with `GET /api/attributes/123`:
```
{
  "data": {
    "id": "123",
    "type": "attributes",
    "relationships": {
      "labels": {
        "links": {
          "self": "http://127.0.0.1:8080/api/attributes/123/relationships/labels",
          "related": "http://127.0.0.1:8080/api/attributes/123/labels"
        }
      }
    },
    "links": {
      "self": "http://127.0.0.1:8080/api/attributes/123"
    }
  }
}
```
The result is as expected. Now I follow the generated `related` link: `GET /api/attributes/123/labels`. I would expect following response:
```
{
  "data": [{
    "id": "9",
    "type": "labels",
    "relationships": {
      "attribute": {
        "links": {
          "self": "http://127.0.0.1:8080/api/labels/9/relationships/attribute",
          "related": "http://127.0.0.1:8080/api/labels/9/attribute"
        }
      }
    },
    "links": {
      "self": "http://127.0.0.1:8080/api/labels/9"
    }
  }]
}
```
Instead I'm getting empty collection:
```
{
  "data": []
}
```
The same empty response I get by following the other `related` link `/api/labels/9/attribute`.

Perhaps I missed something in the documention (I would expect that around this section: https://www.crnk.io/releases/stable/documentation/#_jsonapirelation), maybe some annotation? Or is there actual problem in CRNK core or in used in-memory repository?

**Comment by remmeier** [comment]
I think I see the issue. Not quite sure what the expected behavior should be. Maybe the one insert should be be prohibited.

You do the insert of the relationship like this:

```
   "labels": {
        "data": [{ "type": "labels", "id": "9" }]
      }
```

whereas the model defines:

```
  @JsonApiRelation(mappedBy="attribute")
    private List<Label> labels;
```

which means all the things are "driven" by the attribute field rather than labels. The POST will safe the labels along with the other attributes in the new attributes resource. But it will not trigger an update of the labels resources. Maybe it should be prohibited by making the field "read-only", or implemented manually in the repository. Currently I have the feeling like not automatically triggering an update of opposite resources to follow the REST semantics where a POST/PATCH updates a resource, but nothing else.

**Comment by david-kubecka** [comment]
Ah, so I made a beginners mistake (mixing up the direction of ownership) but as often happens this lead to unexpected usage of the software revealing a bug/undefined behaviour :-)

Regarding the correct behaviour in this case I would at first expect some kind of error being generated. But thinking about it more I wonder what actually happens under the hood of these mapped relations. (Is it actually documented somewhere? I don't see it in https://www.crnk.io/releases/stable/documentation/#_jsonapirelation.) How come that the labels relation is actually stored at all? If I modelled my scenario in relational DB there would be attributes and labels tables with labels having FK to attributes. So when creating new attribute there would actually be no "slot" (column) where to store the relation (assuming you update only the attributes table). Or am I missing something?

**Comment by david-kubecka** [comment]
FYI: I asked probably related question (related by the need of knowing internal details of relationships) here: https://stackoverflow.com/questions/58740359/filtering-bidirectional-relationship-in-crnk

**Comment by remmeier** [comment]
>  I don't see it in https://www.crnk.io/releases/stable/documentation/#_jsonapirelation.
`DEFAULT makes use of FORWARD_OWNER if the relationship makes use of @JsonApiRelationId (see below), lookUp=NONE or lookUp=DEFAULT. FORWARD_OPPOSITE if mappedBy points to another field. CUSTOM if a matching relationship repository is found. For all other scenarios it it expects a custom implementation.`

but it could point into this direction in the mappedBy section a bit above.

yes it should prevent it from being saved. The `mappedBy` is a very recent addition in 3.0, so missed it as well so far.

**Comment by david-kubecka** [comment]
`mappedBy` being undocumented so far is one thing and not such a big problem IMO since it's mostly analogous (I'm guessing here) to the previous `opposite` functionality.

What I woulkd appreciate more in the docs are some examples of the various combinations of `mappedBy`, `lookUp`, `repositoryBehavior`, `@JsonApiRelationId` etc. with the explanation what happens under the hood, ideally expressed with the language of repositories' method calls. I could possibly write that doc myself once I familiarize myself more with crnk.

Btw is there any way how to debug particular GET or POST request? I.e. what gets actually called under the hood? The debug startup events, e.g. `metadata.model.Label.attribute: setting up owner/owner forwarding repository` are certainly useful but I'm still not able to answer my previous question. 

> How come that the labels relation is actually stored at all? If I modelled my scenario in relational DB there would be attributes and labels tables with labels having FK to attributes. So when creating new attribute there would actually be no "slot" (column) where to store the relation (assuming you update only the attributes table).

**Comment by remmeier** [comment]
I pushed a first change with some clarifications. 

Behind the scenes there is SetOppositeStrategy, GetFromOwnerStrategy and ForwardingRelationshipRepository that should improve upon logging.

**Comment by remmeier** [comment]
there is a further section about in-memory data structures in the data chaper now.

**Comment by david-kubecka** [comment]
Thank you, it's much clearer now!

So only think possibly remaining in this issue is to implement a check for wrongly stored relationships. I'm not sure though whether this should be implemented on crnk side. On the one hand it has all the ownership info. On the other the repositories contract maybe is intentionally loose to provide sufficient freedom to custom implementations. I have a feeling, though, that because crnk provides the magic of inclusions and relationship behaviour it should also require stricter contract so that magic can work reliably! But since I'm not an expert I leave the decision upon you.

So either close this issue, unless you want to provide further doc enhancements or in-memory repo fixes, and/or create follow-up for the strict checking of relationships setting.

**Comment by david-kubecka** [comment]
I forgot to mention that I've replicated the steps also on JPA backed resources and the behavior is similarly "strange" (creating an attribute with some labels completely ignores the labels relationship).

**Comment by remmeier** [comment]
yes, JPA in general has exactly the same behavior. The non-owning side are ignored on updates. At least for Hibernate.

have to thinkg about it. one avenue that for sure would be useful is to extend the TestKit to verify the bi-directionality of relationships.

**Comment by david-kubecka** [comment]
I would like to stress that there are two separate aspects. One is the updates which I think is ok to leave as is (perhaps implement some checks on crnk sides).

The other one is filtering the bi-directional relationship through both sides as explained in my comments [here](https://stackoverflow.com/a/58760654/10366910). Basically the arguments goes along the way that if there's a possibility for the repository (even in-memory) to be configured in such a way so that `include` work correctly there should be similar way also for filtering.

**Comment by remmeier** [comment]
>  way so that include work correctly there should be similar way also for filtering

the problem here is that filtering and sorting are way, way harder problems than inclusion. Inclusion is basically just a simple ID to resource lookup (the implementation is not so simple tough...). In contrast, filtering and sorting require detailed knowledge of the data structures and applicable filters. Doing it wrongly or in an inefficient way means spending seconds, minutes and hours on requests for all except small use cases. That is why crnk puts the reponsiblity to the repositories and does not attempt to start doing its own query planning and execution. What do you think?

**Comment by david-kubecka** [comment]
Yes, even from my limited (so far) knowledge of the CRNK framework and the problem area in general I'm aware of the different levels of difficulty of doing simple lookups (inclusion) on one hand, and filtering, sorting etc. on the other hand. I also understand your concerns about performance. Still let me share how I view the problem.

In my opinion a generally usable JSON API framework should support different kinds of repositories (DB based, in-memory, completely custom from various data sources etc.). But the internals of the particular repositories shouldn't leak to public API, i.e. they shouldn't influence the functionality which should be guaranteed by the framework. (What the internals can and will influence is the performance. But that's secondary topic for the sake of this discussion.) In order for the framework to be able to guarantee the prescribed functionality it has to provide a contract by which repositories can declare their capabilities.

And now we get to the main question which is the richness level of such contract. Currently there are already options like `serialize`, `lookUp`, `repositoryBehavior`. I can imagine that there might be much more options here, one of which implies whether/how the repository handles filtering. Or perhaps that could be guided by the already existing `repositoryBehavior`. The docs say
```
FORWARD_OWNER forward any relationship request to the owning resource repository,
```
In my understanding of the problem this
```
GET http://127.0.0.1:8080/api/projects?filter[tasks.id]=1
```
should logically lead into internal "relationship request" and the framework should be able to organize the requests to relevant repositories so that the response is returned as expected. How this is exactly done should depend on the capabilities provided by the particular repository. If the repository is able to perform some kind of join in presence of filter then we are done. If not the framework should internally perform the lookup (similarly as with `include`) and then either do the filtering operation itself or pass the result to some repository method (perhaps directly to `findAll` as second parameter?).

The point is that with the rich contract in place there should be no performance penalty for the more capable repositories (e.g. JPA), while the less capable ones could still achieve the desired functionality albeit with limited performace. But the less capable (e.g. in-memory) repositories aren't expected to hold "big" data so the extra performance impact might not be an issue at all (in contrast to missing filtering functionality).

Does this make any sense? Let me know what you think.

P.S.: The paragraph about `repositoryBehavior` in the doc probably requires some update as well. It references some concepts which no longer exists, e.g. `ImplicitOwnerBasedRelationshipRepository`, `IMPLICIT_GET_OPPOSITE_MODIFY_OWNER`.

**Comment by remmeier** [comment]
Yes, sounds like a good idea. One would have to be a bit cautious, both in terms of implementation and not increase the complexity, but there likely is room for improment/more automation.

> I can imagine that there might be much more options here, one of which implies whether/how the repository handles filtering. 

that is already in place, there is @JsonApiField which offers `sortable`, `filterable`, `patchable`, etc. The next thing that would come into my mind is a `joinable` that specifies whether sorting and filtering can be applied along that path. This in turn could allow one or the other better optimization.
